packages/oruga/src/composables/useInputHandler.ts

```diff
@@ -203,6 +220,85 @@ export function useInputHandler(
+        // Because of that, this will only work properly if the `required` attributes of all radio
+        // buttons in the group are synchronized with each other, which is likely anyway.
+        // (We're also expecting the use of radio buttons with our default validation message handling
+        // to be fairly uncommon because the overall visual experience is clunky with such a configuration.)
+        const onAttributeChange = (): void => {
+            if (!isValid.value) checkHtml5Validity();
+        };
+        let validationAttributeObserver: MutationObserver | null = null;
+        let ancestorMutationObserver: MutationObserver | null = null;
+        watch(
+            [maybeElement, isValid, () => props.useHtml5Validation],
+            (data) => {
+                // Not using destructuring assignment because browser support is just a little too weak at the moment
+                const el = data[0];
+                const valid = data[1];
+                const useValidation = data[2];
+
+                // Clean up previous state.
+                if (validationAttributeObserver != null) {
+                    // Process any pending events.
+                    if (validationAttributeObserver.takeRecords().length > 0) {
+                        onAttributeChange();
+                    }
+                    validationAttributeObserver.disconnect();
+                }
+                if (ancestorMutationObserver != null) {
+                    if (ancestorMutationObserver.takeRecords().length > 0) {
+                        onAttributeChange();
+                    }
+                    ancestorMutationObserver.disconnect();
+                }
+
+                if (el == undefined || valid || !useValidation) {
+                    return;
+                }
+
+                if (validationAttributeObserver == null) {
+                    validationAttributeObserver = new MutationObserver(
+                        onAttributeChange,
+                    );
+                }
+                validationAttributeObserver.observe(el, {
+                    attributeFilter: constraintValidationAttributes,
+                });
+
+                // Note that this doesn't react to changes in the list of ancestors.
+                // Based on testing, Vue seems to rarely, if ever, re-parent DOM nodes;
+                // it generally prefers to create new ones under the new parent.
+                // That means this simpler solution is likely good enough for now.
+                let ancestor: Node | null = el;
+                while ((ancestor = ancestor.parentNode)) {
+                    // Form controls can be disabled by their ancestor fieldsets.
+                    if (ancestor instanceof HTMLFieldSetElement) {
+                        if (ancestorMutationObserver == null) {
+                            ancestorMutationObserver = new MutationObserver(
```

Comment:
I haven't worked much with observers yet. Is it really necessary to have multiple observers, or is it possible to have one observer for multiple targets?

Reply:
One observer can be attached to multiple targets; in fact, `ancestorMutationObserver` is tied to several DOM elements (every ancestor of the form control). Now that you mention it, though, I do think we can use the same observer to watch both the form element (and all its relevant attributes) and the ancestors (for just the `disabled` attribute). The two-observer strategy is partially a hold-over from a much more elaborate solution I was working on, which tried to detect cases where the form control got re-parented into another part of the DOM. Since Vue doesn't seem to re-parent nodes in typical use cases, I abandoned that approach to avoid the complexity.

I've pushed up [a patch](https://github.com/oruga-ui/oruga/pull/930/commits/dcd62557f5fa80eb36c7b862e676771278aeeef8).